adapter/internal/ga/client.go

```diff
@@ -288,13 +290,98 @@ func addAPIToChannel(resp *discovery.DiscoveryResponse) {
 			IsDeployEvent:    false,
 			OrganizationUUID: gaAPI.OrganizationUUID,
 			RevisionUUID:     gaAPI.RevisionUUID,
+			DeployedEnv:      gaAPI.DeployedEnv,
 		}
 		GAAPIChannel <- event
 		delete(apiRevisionMap, apiEntry)
 		logger.LoggerGA.Infof("API Undeploy event is added to the channel. : %s", apiEntry)
 	}
 }
 
+func addAPIWithEnvToChannel(resp *discovery.DiscoveryResponse) {
+
+	// removedAPIEnvMap is used To keep track of the APIs needs to be deleted.
+	var removedAPIEnvMap = make(map[string]map[string]*ga_model.Api)
+
+	if !isFirstResponse {
+		for k, v := range apiEnvRevisionMap {
+			cp := make(map[string]*ga_model.Api)
+			for l, w := range v {
+				cp[l] = w
+			}
+			removedAPIEnvMap[k] = cp
+		}
+	}
+
+	var startupAPIEventArray []*APIEvent
+	// Even if there are no resources available within ga, an empty but non-nil array would be set
+	startupAPIEventArray = make([]*APIEvent, 0)
+	for _, res := range resp.Resources {
+		api := &ga_model.Api{}
+		err := ptypes.UnmarshalAny(res, api)
+
+		if err != nil {
+			logger.LoggerGA.Errorf("Error while unmarshalling: %s\n", err.Error())
+			continue
+		}
+
+		currentGAAPI, apiFound := apiEnvRevisionMap[api.ApiUUID][api.DeployedEnv]
+		if apiFound {
+			if currentGAAPI.RevisionUUID == api.RevisionUUID {
+				logger.LoggerGA.Debugf("Current GA API revision ID and API event revision ID: %s in "+
+					"environment: %s are equal", currentGAAPI.RevisionUUID, currentGAAPI.DeployedEnv)
+				delete(removedAPIEnvMap[api.ApiUUID], api.DeployedEnv)
+				continue
+			}
+		}
+		event := APIEvent{
+			APIUUID:          api.ApiUUID,
+			RevisionUUID:     api.RevisionUUID,
+			IsDeployEvent:    true,
+			OrganizationUUID: api.OrganizationUUID,
+			DeployedEnv:      api.DeployedEnv,
+		}
+
+		// If it is the first response, the GA would not send it via the channel. Rather
+		// it appends to an array and let the apis_fetcher collect those data.
+		if isFirstResponse {
+			startupAPIEventArray = append(startupAPIEventArray, &event)
+		} else {
+			GAAPIChannel <- event
+		}
+		if _, found := apiEnvRevisionMap[api.ApiUUID]; !found {
+			apiEnvRevisionMap[api.ApiUUID] = make(map[string]*ga_model.Api)
+		}
+		apiEnvRevisionMap[api.ApiUUID][api.DeployedEnv] = api
+		logger.LoggerGA.Infof("API Deploy event is added to the channel. %s : %s : %s", api.ApiUUID,
+			api.RevisionUUID, api.DeployedEnv)
+	}
+
+	// If it is the first response, it does not contain any remove events.
+	if isFirstResponse {
+		initialAPIEventArray = startupAPIEventArray
+		isFirstResponse = false
+		return
+	}
+
+	for apiID, envAPI := range removedAPIEnvMap {
+		for envName, gaAPI := range envAPI {
+			event := APIEvent{
+				APIUUID:          apiID,
+				IsDeployEvent:    false,
+				OrganizationUUID: gaAPI.OrganizationUUID,
+				RevisionUUID:     gaAPI.RevisionUUID,
+				DeployedEnv:      envName,
+			}
+			GAAPIChannel <- event
+			delete(apiEnvRevisionMap[apiID], envName)
```

Comment:
Let's avoid possible NPE

Reply:
added a check prior to deletion https://github.com/wso2/product-microgateway/commit/ae2c06882353bda7a59101381ef3e82f88667a21